tests/topotests/lib/topotest.py

```diff
@@ -66,144 +67,140 @@ def has_errors(self):
         "Returns True if there were errors, otherwise False."
         return len(self.errors) > 0
 
+    def gen_report(self):
+        headline = ["Generated JSON diff error report:", ""]
+        return headline + self.errors
+
     def __str__(self):
-        return "\n".join(self.errors)
+        return "Generated JSON diff error report:\n\n\n" + "\n".join(self.errors) + "\n\n"
 
 
-def json_diff(d1, d2):
-    """
-    Returns a string with the difference between JSON data.
-    """
-    json_format_opts = {
-        "indent": 4,
-        "sort_keys": True,
-    }
-    dstr1 = json.dumps(d1, **json_format_opts)
-    dstr2 = json.dumps(d2, **json_format_opts)
-    return difflines(dstr2, dstr1, title1="Expected value", title2="Current value", n=0)
-
-
-def _json_list_cmp(list1, list2, parent, result):
-    "Handles list type entries."
-    # Check second list2 type
-    if not isinstance(list1, type([])) or not isinstance(list2, type([])):
-        result.add_error(
-            "{} has different type than expected ".format(parent)
-            + "(have {}, expected {}):\n{}".format(
-                type(list1), type(list2), json_diff(list1, list2)
-            )
-        )
-        return
-
-    # Check list size
-    if len(list2) > len(list1):
-        result.add_error(
-            "{} too few items ".format(parent)
-            + "(have {}, expected {}:\n {})".format(
-                len(list1), len(list2), json_diff(list1, list2)
-            )
-        )
-        return
-
-    # List all unmatched items errors
-    unmatched = []
-    for expected in list2:
-        matched = False
-        for value in list1:
-            if json_cmp({"json": value}, {"json": expected}) is None:
-                matched = True
-                break
-
-        if not matched:
-            unmatched.append(expected)
-
-    # If there are unmatched items, error out.
-    if unmatched:
-        result.add_error(
-            "{} value is different (\n{})".format(parent, json_diff(list1, list2))
-        )
+def gen_json_diff_report(d1, d2, exact=False, path="> $", acc=(0 , "")):
```

Comment:
Would mind adding a small description to this function? Maybe tell the reader this is an internal function.

Other than that I think this is mostly it, the new formatter might want to change some long lines here.

Reply:
Done.

I used the python string formatter whereever it makes sense and re-formatted the file using `black`.